src/System.Memory/src/System/Buffers/ReadOnlyBuffer.cs

```diff
@@ -0,0 +1,417 @@
+
+        /// <summary>
+        /// Forms a slice out of the given <see cref="ReadOnlyBuffer{T}"/>, beginning at '<paramref name="start"/>, and is at most <paramref name="length"/> bytes
+        /// </summary>
+        /// <param name="start">The starting (inclusive) <see cref="SequencePosition"/> at which to begin this slice.</param>
+        /// <param name="length">The length of the slice</param>
+        public ReadOnlyBuffer<T> Slice(SequencePosition start, int length)
+        {
+            BoundsCheck(BufferEnd, start);
+
+            var end = Seek(start, BufferEnd, length, false);
+
+            return SliceImpl(start, end);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given <see cref="ReadOnlyBuffer{T}"/>, beginning at <paramref name="start"/>, ending at <paramref name="end"/> (inclusive).
+        /// </summary>
+        /// <param name="start">The starting (inclusive) <see cref="SequencePosition"/> at which to begin this slice.</param>
+        /// <param name="end">The ending (inclusive) <see cref="SequencePosition"/> of the slice</param>
+        public ReadOnlyBuffer<T> Slice(SequencePosition start, SequencePosition end)
+        {
+            BoundsCheck(BufferEnd, end);
+            BoundsCheck(end, start);
+
+            return SliceImpl(start, end);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given <see cref="ReadOnlyBuffer{T}"/>, beginning at <paramref name="start"/>, ending at the existing <see cref="ReadOnlyBuffer{T}"/>'s end.
+        /// </summary>
+        /// <param name="start">The starting (inclusive) <see cref="SequencePosition"/> at which to begin this slice.</param>
+        public ReadOnlyBuffer<T> Slice(SequencePosition start)
+        {
+            BoundsCheck(BufferEnd, start);
+
+            return SliceImpl(start, BufferEnd);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given <see cref="ReadOnlyBuffer{T}"/>, beginning at <paramref name="offset"/>, ending at the existing <see cref="ReadOnlyBuffer{T}"/>'s end.
+        /// </summary>
+        /// <param name="offset">The start index at which to begin this slice.</param>
+        public ReadOnlyBuffer<T> Slice(long offset)
+        {
+            if (offset == 0) return this;
+
+            var begin = Seek(BufferStart, BufferEnd, offset, false);
+            return SliceImpl(begin, BufferEnd);
+        }
+
+        /// <summary>
+        /// Copy the <see cref="ReadOnlyBuffer{T}"/> to the specified <see cref="Span{Byte}"/>.
+        /// </summary>
+        /// <param name="destination">The destination <see cref="Span{Byte}"/>.</param>
+        public void CopyTo(Span<T> destination)
+        {
+            if (Length > destination.Length)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.destination);
+            }
+
+            foreach (var buffer in this)
+            {
+                buffer.Span.CopyTo(destination);
```

Comment:
Do we like the callstack/excpetion that might propagate from copyto when the destination is too small?